### PR TITLE
feat(nextjs): can set compiler when setting up `nxComponentTestingPreset`

### LINKED ISSUE #19131: Cypress Code Coverage and nextjs
- [X] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
I would like Cypress Code Coverage to work in a *next* project when using `nxComponentTestingPreset`.  Currently it works if you are using a *react* project and importing `nxComponentTestingPreset` from `@nx/react/plugins/component-testing`, but tests using `NavLink` fail because they are missing `process`.  

There is a slight difference in `nxComponentTestingPreset` between *next* and *react*.  They use a different `compiler`.  *next* uses **swc** and *react* uses **babel**.  If I change my local code in `next/plugins/compoent-testing.js` to use **babel** instead of **swc** I get proper code coverage results in *next*.

## Motivation
Having Cypress Code Coverage working in a *next* project would be nice, and something the company I work for is striving to implement coverage requirements.

## Suggested Implementation
I imagine adding a `compiler` property to `NxComponentTestingOptions` and if it is set use that property here [compiler: 'swc'](https://github.com/nrwl/nx/blob/f6398837869cc3d8497487c427a16af1f821c46d/packages/next/plugins/component-testing.ts#L100)

I would change 
```typescript
compiler: 'swc',
```

to

```typescript
compiler: options?.compiler || 'swc',
```

And add `compiler` as an optional property on `NxComponentTestingOptions`.

## Alternate Implementations
Just change it to be `compiler: 'babel'` so it matches the *react* configuration.

### PR DESCRIPTION
**Pull request by erenken**
Currently *nextjs* only uses the *swa* compiler in Cypress component testing.

To use @cypress/code-coverage you need to use *babel*, you can now select which compiler to use.

closed #19131

## Current Behavior
Cypress Code Coverage doesn't work in a *next* project when using `nxComponentTestingPreset` from `@nx/next/plugins/component-testing`. Currently it works if you are using a *react* project and importing `nxComponentTestingPreset` from `@nx/react/plugins/component-testing`, but tests using NavLink fail because they are missing process with using this preset.

## Expected Behavior
I would expect Cypress Code Coverage to work in a *next* and *react* project.

## Related Issue(s)

Fixes #19131 

Unable to run test on `devcontainer`, so that was updated to use latest version.

### CODE REVIEW

**Comment by mandarini** [comment]
Hey @erenken thanks for this PR! Can you please add some more tests around this?

**Comment by mandarini** [comment]
@erenken changes look good, we just need a few tests around this and the pipelines to pass! :D Just ping me when you're done and I will merge! Thank you for this PR!

**Comment by erenken** [comment]
> @erenken changes look good, we just need a few tests around this and the pipelines to pass! :D Just ping me when you're done and I will merge! Thank you for this PR!

Are there any existing tests for the `nxComponentTestingPreset` method?  I was trying to find them, and I couldn't find anything that executes the code.  The only tests that have `nxComponentTestingPreset` in them are all doing contains or snapshot comparisons.  

Just adding the line:

`const result = nxComponentTestingPreset('myFile.json', {});`

Produces this error:

```typescript
    /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /workspaces/nx/packages/nx/src/native/nx.linux-x64-gnu.node)

      180 |           try {
      181 |             if (localFileExisted) {
    > 182 |               nativeBinding = require('./nx.linux-x64-gnu.node')
          |                               ^
      183 |             } else {
      184 |               nativeBinding = require('@nx/nx-linux-x64-gnu')
      185 |             }

      at CompileFunctionRuntime._loadModule (../../node_modules/.pnpm/jest-runtime@29.5.0/node_modules/jest-runtime/build/index.js:1009:29)
      at Object.<anonymous> (../nx/src/native/index.js:182:31)
      at Object.<anonymous> (../nx/src/hasher/task-hasher.ts:23:1)
      at Object.<anonymous> (../nx/src/hasher/hash-task.ts:4:1)
      at Object.<anonymous> (../nx/src/tasks-runner/task-orchestrator.ts:21:1)
      at Object.<anonymous> (../nx/src/tasks-runner/default-tasks-runner.ts:2:1)
      at Object.<anonymous> (../devkit/nx-reexports-pre16.ts:209:1)
      at Object.<anonymous> (../devkit/index.js:21:20)
      at Object.<anonymous> (../cypress/src/utils/ct-helpers.ts:2:1)
      at Object.<anonymous> (plugins/component-testing.ts:1:1)
      at Object.<anonymous> (plugins/component-testing.spec.ts:1:1)
```

When I comment out that line no errors and the tests pass.  Any recommendations?

**Comment by erenken** [comment]
Never mind, I got it all working after updating the devcontainer.  Pushing tests soon.

**Comment by mandarini** [comment]
Thank you very much for this PR @erenken !

**Review by barbados-clemens — CHANGES_REQUESTED** [review]
Thanks for your PR. There isn't a need to add those deps to the package.json since the tests for component testing should be in the e2e projects where they will be installed at run time in the testing project vs the actual nx repo. 

thanks for updating the dev container.

File: packages/next/project.json [hunk 1]
```diff
@@ -59,6 +59,12 @@
...
       "command": "node ./scripts/add-dependency-to-build.js next @nrwl/next"
     },
-    "lint": {}
+    "lint": {},
+    "component-test": {
+      "executor": "@nx/cypress:cypress",
+      "options": {
+        "cypressConfig": "cypress-component-test.config.ts"
+      }
+    }
```
[the remark this answers is not in the document]
Comment (erenken) [reply packages/next/project.json]: This has been removed.

File: package.json [hunk 2]
```diff
@@ -149,6 +149,7 @@
     "conventional-changelog-cli": "^2.0.23",
     "convert-source-map": "^2.0.0",
     "copy-webpack-plugin": "^10.2.4",
+    "css-loader": "^6.8.1",
```
[the remark this answers is not in the document]
Comment (erenken) [reply package.json:154]: This has been removed.

File: package.json [hunk 3]
```diff
@@ -237,6 +238,7 @@
     "parse5": "4.0.0",
     "postcss": "8.4.19",
     "postcss-import": "~14.1.0",
+    "postcss-loader": "^7.3.3",
```
[the remark this answers is not in the document]
Comment (erenken) [reply package.json:243]: This has been removed.

File: packages/next/plugins/component-testing.spec.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import { NxComponentTestingOptions } from '@nx/cypress/plugins/cypress-preset';
+import { nxComponentTestingPreset } from './component-testing';
+
+describe('nxComponentTestingPreset', () => {
```
[the remark this answers is not in the document]
Comment (erenken) [reply packages/next/plugins/component-testing.spec.ts]: e2e tests have been added, and are all passing, well at least locally :)

File: packages/react/plugins/component-testing/index.ts [hunk 5]
```diff
@@ -160,7 +160,7 @@ export function nxComponentTestingPreset(
     const { buildBaseWebpackConfig } = require('./webpack-fallback');
     webpackConfig = buildBaseWebpackConfig({
       tsConfigPath: findTsConfig(normalizedProjectRootPath),
-      compiler: 'babel',
+      compiler: options?.compiler || 'babel',
```
[the remark this answers is not in the document]
Comment (erenken) [reply packages/react/plugins/component-testing/index.ts:163]: Done

**Comment by erenken** [comment]
FYI ci/circleci: main-linux is giving this error:

`Error when connecting to Nx Cloud. Code: 500. Error: unknown error`

There appears to be a problem with the environment.

**Comment by erenken** [comment]
@mandarini I got all the changes that you and @barbados-clemens wanted.  I removed the packages I added to packages.json and created E2E test, and everything is passing.  If I missed something or you need more, please let me know.  I would like to get this merged in so we can get code coverage working in our project.

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -3,7 +3,7 @@
 {
   "name": "NxDevContainer",
   // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
-  "image": "mcr.microsoft.com/devcontainers/typescript-node:0-18",
```
Comment (barbados-clemens) [inline .devcontainer/devcontainer.json:6]: nitpick, let's not have this change in this PR, also probs shouldn't be latest as could end up using version that isn't fully supported yet. 

Mainly just doing want these changes to end up in PRs that aren't fully related. Sorry didn't catch this initially.

File: pnpm-lock.yaml [hunk 7]
```diff
@@ -533,6 +533,9 @@ devDependencies:
   copy-webpack-plugin:
     specifier: ^10.2.4
     version: 10.2.4(webpack@5.88.0)
+  css-loader:
```
Comment (barbados-clemens) [inline pnpm-lock.yaml:536]: should revert these changes from lock file and package.json.

**Review by barbados-clemens — COMMENTED** [review]
looks like the package.json changes are still there can you revert those real quick?

File: .devcontainer/devcontainer.json (hunk 6 above)
Comment (erenken) [reply .devcontainer/devcontainer.json:6]: Not changing it now, but my thought would be to change it to `0-20`

Changes made.

File: pnpm-lock.yaml (hunk 7 above)
Comment (erenken) [reply pnpm-lock.yaml:536]: Sorry, I thought I got all of them.  I just revered the files to what was in master.

Comment (erenken) [reply pnpm-lock.yaml]: OK, sorry for all of the commits, was having issues with my branches being out of sync.  package and pnpm-lock are both revered.  We should be good now.

**Comment by erenken** [comment]
Look like the `ci/circleci: main-linux` needs to be re-ran it is giving the error:

```
Failed to install browsers
Error: Failed to install chrome
```

**Comment by erenken** [comment]
I know it looks like the main-linux failed, but I think it just needs to be re-ran.  I have found that these pipelines are very temperamental and the only way to re-run them is to commit a change or update from master.  It was working before the last update from master.

**Review by barbados-clemens — APPROVED** [review]

**Review by mandarini — APPROVED** [review]
